Change BucketPrefix methods to BucketName

mrambacher opened the pull request:
The XXXGetBucketPrefix methods were changed to XXXGetBucketName.  Whereas the old methods would return a name and the "GetBucketName" methods (of aws_file) would prepend "rockset." to the name, the new methods return the full name.  

Additionally, environment variables were added/changed to set the PREFIX (to use a value instead of "rockset.").

The impetus for this change was because we cannot specify the name of the buckets in our environment to be prefixed by "rockset.".  Because the GetBucketName methods were not part of a class, it was hard to override the settings.  Additionally, this change will make it easier for other cloud environments to use the same methodology in the future.

ari-e commented on Makefile:
```diff
@@ -308,7 +308,7 @@ endif
 
 WARNING_FLAGS = -W -Wextra -Wall -Wsign-compare -Wshadow \
-  -Wunused-parameter
+  -Wno-unused-parameter
```
Looks like leftover debugging/development code. Revert

ari-e commented on cloud/aws/aws_env.h:
```diff
@@ -273,9 +274,7 @@ class AwsEnv : public CloudEnvImpl {
-  static Status GetTestCredentials(std::string* aws_access_key_id,
-                                   std::string* aws_secret_access_key,
-                                   std::string* region);
+  static Status GetTestCredentials(std::string *aws_access_key_id, std::string *aws_secret_access_key, std::string* region);
```
Revert; doesn't adhere to our style guide

ari-e commented on cloud/aws/aws_file.h:
```diff
@@ -41,27 +41,6 @@
-}
-
-}  // namespace
-
```
The one problem with removing this is that now there are a lot of lines like `Aws::String(X.c_str(), X.size())` that you've introduced which are a bit clunky. Maybe add a helper function here like `AWS::String toAwsString(const std::string& s);` to clean it up

ari-e commented on env/env_basic_test.cc:
```diff
@@ -91,6 +91,7 @@ static std::unique_ptr<Env> mock_env(new MockEnv(Env::Default()));
 INSTANTIATE_TEST_CASE_P(MockEnv, EnvBasicTestWithParam,
                         ::testing::Values(mock_env.get()));
 #ifndef ROCKSDB_LITE
+
```
?

ari-e commented on include/rocksdb/cloud/cloud_env_options.h:
```diff
@@ -30,6 +30,8 @@ enum LogType : unsigned char {
 };
 
+
+
```
?

ari-e commented on tools/db_bench_tool.cc:
```diff
@@ -1062,7 +1062,7 @@ enum RepFactory {
 // create Factory for creating S3 Envs
 #ifdef USE_AWS
-rocksdb::Env* CreateAwsEnv(const std::string& dbpath,
+rocksdb::Env* CreateAwsEnv(const std::string& ,
```
Typo?

mrambacher replied:
The dbpath option is not used, hence why the name is deleted.  This will also lead to errors/warnings if the unused flag is turned on.  How do you want to leave it?

igorcanadi commented on cloud/aws/aws_env.cc:
```diff
@@ -221,38 +222,86 @@ Aws::S3::Model::HeadObjectOutcome AwsS3ClientWrapper::HeadObject(
+  
+static std::string GetFullBucketName(const std::string & name, const char *bucket) {
+  std::string full_name;
+  if (! GetNameFromEnvironment(bucket, "ROCKSDB_CLOUD_BUCKET_NAME", &full_name)) {
```
I would prefer not use environment variable to control a thing such as bucket name.

What do you think about the following: add an option in CloudEnvOptions `bucket_name_add_rockset_prefix`, defaults to true for backwards compatibility. You can then set it to `false` in your environment.

mrambacher replied:
We *really* want to control all of RocksSet without using hard-coded parameters.  We hope to use RocksDB as a bundle plugged into software we do not compile or own.  That is the point of most of our changes.

In the future we hope to be able to control things like bucket name via INI files (as well as possibly the environment).  But the environment is our only option at the moment.

What is the object to controlling some of the parameters through the environment?

igorcanadi replied:
@mrambacher I agree with you that you want to fully control the bucket path. What I'm saying is in line with that -- a parameter given to `CloudEnv::NewAwsEnv()` needs to be full bucket name and RocksDB-cloud should not add *any* prefix to it. Right now it is adding a prefix that is controlled by the environment variable, but I think it should not even do that. Don't add any prefix, period.

Do you agree with this?

However, we'll unfortunately need to keep the backwards compatibility, which adds `rockset.` as a prefix to the bucket name. Let's control this behavior using `CloudEnvOptions`. Does this makes sense?

> But the environment is our only option at the moment.

It is fine if your code controls the bucket name using the environment variable. But you should read the environment variable in the binary and provide the bucket_name value in `CloudEnv::NewAwsEnv()`.

mrambacher replied:
The problem is that I do not necessarily have control of the binary going forward.   I want to integrate the Cloud code into Rocksandra or MyRocks.  I do not want to modify the code of those products to set additional parameters to specify bucket names or prefixes.  Ideally, I want to get rid of the options to NewAwsEnv, not add more.

I also believe there are probably other examples of code trying to replace the "rockset." prefix already.  In db_test_util, "dbtest." is prepended to the name of the bucket.  I would assume that the code would think buckets are named by this parameter and not rockset.dbtest."

If you really want, I can add a "prefix" to the CloudEnvOptions.  I can also add the other bucket parameters to CloudEnvOptions at the same time.  However, I still need a means of setting (or at least initializing) them to some other value via the environment.

igorcanadi replied:
I think the rule of thumb should be that all configuration happens through `CloudEnvOptions`. We can initialize `CloudEnvOptions` through either environment variables or .ini files or whatever else. But once initialized, we should pass `CloudEnvOptions` to `NewAwsEnv()` and all behavior of RocksDB-Cloud should be evident by inspecting contents of `CloudEnvOptions`.

Does this make sense?

> If you really want, I can add a "prefix" to the CloudEnvOptions. I can also add the other bucket parameters to CloudEnvOptions at the same time.

Yes, let's do that.

> However, I still need a means of setting (or at least initializing) them to some other value via the environment.

Feel free to introduce a new function `InitializeCloudEnvFromEnvironment` or even `IntroduceCloudEnvFromConfigFile`. I do not know how we'll invoke this function from MyRocks, but whatever we do, we should have a clean separation. Would that work for you?

igorcanadi replied:
> I also believe there are probably other examples of code trying to replace the "rockset." prefix already. In db_test_util, "dbtest." is prepended to the name of the bucket. I would assume that the code would think buckets are named by this parameter and not rockset.dbtest."

Yes, I do not like this transparent adding of `rockset.` prefix. If we migrate away from it, we will break backwards compatibility, though :(

mrambacher replied:
I pushed out another changeset to this PR today.  Because of its where we are in the review, I did not merge it into the other change set.  In this changeset, I added a BucketOptions class that contains the settings (bucket name, object prefix, region) for a bucket.  There are now two of these (source, dest) in a CloudEnvOptions.  When created, the bucket settings are initialized from the environment.  You can also explicitly override them via Set methods.  There is also an "Initialize" method that sets the values to the input ones if the input ones are not blank.

I also added another NewAwsEnv signature that has all of these values in the options. 

Please take a look at the update.  I have tested db_cloud_test and everything passes here.

igorcanadi replied:
Overall I like the change. Two general comments:
1) `CloudEnv::NewAwsEnv` now takes bucket config from both the function arguments and also from `CloudEnvOptions`. Specifying one thing in two different ways is not a good API design. We should remove the function arguments. This will break the backwards compatibility, so we need to make sure to specify the change in release notes (HISTORY.md).
2) I am still not on board with the default constructor of `BucketOptions` reading environment variables. I would introduce `BucketOptions initializeSourceBucketFromEnviroment()` and `BucketOptions intializeDestinationBucketFromEnvironment()`. Given the today's design of RocksDB-Cloud, we will need to have a different initialization path that we include in MyRocks and Rocksandra. We can call these two functions from there (in the future).

igorcanadi commented on cloud/aws/aws_env.cc:
```diff
@@ -221,38 +222,86 @@ Aws::S3::Model::HeadObjectOutcome AwsS3ClientWrapper::HeadObject(
+  src_bucket_name_   = GetFullBucketName(src_bucket_name, "ROCKSDB_CLOUD_SRC_BUCKET_NAME");
+  src_object_path_ = trim(src_object_path);
+  dest_bucket_name_  = GetFullBucketName(dest_bucket_name, "ROCKSDB_CLOUD_DEST_BUCKET_NAME");
+  dest_object_path_ = trim(dest_object_path);
```
Feel free to remove the trim business here.

mrambacher replied:
The values were trimmed before.  I put the trim in here because of that.  Am removing per your suggestion.

igorcanadi commented:
@mrambacher I like this change, thanks for doing this! Sorry about adding `rockset.` as a prefix by default earlier. Please see my inline comment.

> Also, please go ahead and squash your 27 commits down to just 1 for a cleaner git history for when we can get this approved and merged.

@ari-e Github now allows us to squash the commits before merging ;)

mrambacher commented on Makefile:
```diff
@@ -308,7 +308,7 @@ endif
 
 WARNING_FLAGS = -W -Wextra -Wall -Wsign-compare -Wshadow \
-  -Wunused-parameter
+  -Wno-unused-parameter
```
I reverted this back.  However, this makes the RockSet code compile differently than RocksDB, which seems like a dangerous precendent.

igorcanadi replied:
I agree with this change, but let's do this in a separate PR.

mrambacher commented:
I am fine with removing the extra constructor. I rather head to a
constructor with no args.

As for the environment and extra methods, those methods would not help.
First, I have no way to call them when I don't control the executable.
Second, how do you make all of the tests work? They would need to call it
both ways, which kind of defeats the purpose. If the values are not set,
everything acts as it does today.

On Fri, Jun 21, 2019, 12:55 Igor Canadi <notifications@github.com> wrote:

> *@igorcanadi* commented on this pull request.
> ------------------------------
>
> In cloud/aws/aws_env.cc
> <https://github.com/rockset/rocksdb-cloud/pull/58#discussion_r296315907>:
>
> > +static bool GetNameFromEnvironment(const char *name, const char *alt, std::string * result) {
> +  char *value = getenv(name);
> +  if (value == nullptr && alt != nullptr) {
> +    value = getenv(alt);
> +  }
> +  if (value != nullptr) {
> +    result->assign(value);
> +    return true;
> +  } else {
> +    return false;
> +  }
> +}
> +
> +static std::string GetFullBucketName(const std::string & name, const char *bucket) {
> +  std::string full_name;
> +  if (! GetNameFromEnvironment(bucket, "ROCKSDB_CLOUD_BUCKET_NAME", &full_name)) {
>
> Overall I like the change. Two general comments:
>
>    1. CloudEnv::NewAwsEnv now takes bucket config from both the function
>    arguments and also from CloudEnvOptions. Specifying one thing in two
>    different ways is not a good API design. We should remove the function
>    arguments. This will break the backwards compatibility, so we need to make
>    sure to specify the change in release notes (HISTORY.md).
>    2. I am still not on board with the default constructor of
>    BucketOptions reading environment variables. I would introduce BucketOptions
>    initializeSourceBucketFromEnviroment() and BucketOptions
>    intializeDestinationBucketFromEnvironment(). Given the today's design
>    of RocksDB-Cloud, we will need to have a different initialization path that
>    we include in MyRocks and Rocksandra. We can call these two functions from
>    there (in the future).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rockset/rocksdb-cloud/pull/58?email_source=notifications&email_token=AABCTBEPMXU7ZNE4FSZA423P3UBZ5A5CNFSM4HYDI4RKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4JWNMA#discussion_r296315907>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABCTBCO6O2TMGXEVMK7BO3P3UBZ5ANCNFSM4HYDI4RA>
> .
>

igorcanadi commented:
> First, I have no way to call them when I don't control the executable.

I'm not completely sure how we'll make RocksDB cloud without controlling the executable. I'm assuming we'll intercept RocksDB's DB constructor and call some RocksDB-cloud-specific code there. Is that right? Why don't we initialize `BucketOptions` in that piece of code? Some code somewhere will have to invoke `NewAwsEnv`. That some code should also initialize `BucketOptions` from environment variables. But we should leave it to the users of the library to decide whether to use environment variables or not.

> Second, how do you make all of the tests work? They would need to call it
both ways, which kind of defeats the purpose. If the values are not set,
everything acts as it does today.

I don't fully understand the problem with tests. Tests can just use the current environment variable to initialize `BucketOptions`, just like they do now.

mrambacher commented:
I am working on being able to use RocksDB cloud without controlling the
executable.  By a combination of using dynamic libraries, the object
registry, and configuration files, I am pretty close to being able to load
RocksDB Cloud without changing the calling executable.  I am working
through changes in RocksDB to allow this to happen, and working on making
the corresponding changes to RocksDB Cloud as well.

As it stands now, I cannot run the RocksDB cloud tests in my environment
because it using appends "rockset." to all of the bucket names.  Without
environment variables, I cannot do anything.

And there is a precedent for environment variables for RocksDB Cloud
configurations.  They are used for credentials and AWS settings already.

In the long run, I think there should be three ways to set the RocksDB
configuration properties:
1.  Through the API/code
2.  Through the environment (for "more important" settings)
3.  Through configuration files

This change implements #2 without changing the behavior of anything.
Existing code should continue to work exactly as it does today.



On Fri, Jun 21, 2019 at 1:36 PM Igor Canadi <notifications@github.com>
wrote:

> First, I have no way to call them when I don't control the executable.
>
> I'm not completely sure how we'll make RocksDB cloud without controlling
> the executable. I'm assuming we'll intercept RocksDB's DB constructor and
> call some RocksDB-cloud-specific code there. Is that right? Why don't we
> initialize BucketOptions in that piece of code? Some code somewhere will
> have to invoke NewAwsEnv. That some code should also initialize
> BucketOptions from environment variables. But we should leave it to the
> users of the library to decide whether to use environment variables or not.
>
> Second, how do you make all of the tests work? They would need to call it
> both ways, which kind of defeats the purpose. If the values are not set,
> everything acts as it does today.
>
> I don't fully understand the problem with tests. Tests can just use the
> current environment variable to initialize BucketOptions, just like they
> do now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rockset/rocksdb-cloud/pull/58?email_source=notifications&email_token=AABCTBGP257OL7IQRTGHU4TP3UGRNA5CNFSM4HYDI4RKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYJDFYA#issuecomment-504509152>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABCTBGOEJOCB7ZVUYWG2ZLP3UGRNANCNFSM4HYDI4RA>
> .
>

igorcanadi commented:
> I am working on being able to use RocksDB cloud without controlling the
executable.  By a combination of using dynamic libraries, the object
registry, and configuration files, I am pretty close to being able to load
RocksDB Cloud without changing the calling executable.  I am working
through changes in RocksDB to allow this to happen, and working on making
the corresponding changes to RocksDB Cloud as well.

Can you explain briefly where will you invoke `CloudEnv::NewAwsEnv()` in this setup?

> And there is a precedent for environment variables for RocksDB Cloud
configurations.  They are used for credentials and AWS settings already.

Environment variables are fine for credentials, I agree. They are not fine for an option that controls the path to which RocksDB Cloud syncs the data. I am happy to have hooks that lets you initialize config through environment variables, but they should not be invoked by default by the library.

mrambacher commented:
The idea is that there will be somewhere in a RocksDB INI file a field that
says use "AWS" as your environment and that will case the NewAwsEnv to be
used.  There is an existing ObjectRegistry in RocksDB.  Imagine how you
would use it to initialize RocksDB Cloud.

(I realize that this does not work completely at the moment for RocksDB
Cloud as you require a custom DB::Open call.  But that problem is solvable
too with a better pluggable mechanism for DB::Open).



On Mon, Jun 24, 2019 at 6:18 AM Igor Canadi <notifications@github.com>
wrote:

> I am working on being able to use RocksDB cloud without controlling the
> executable. By a combination of using dynamic libraries, the object
> registry, and configuration files, I am pretty close to being able to load
> RocksDB Cloud without changing the calling executable. I am working
> through changes in RocksDB to allow this to happen, and working on making
> the corresponding changes to RocksDB Cloud as well.
>
> Can you explain briefly where will you invoke CloudEnv::NewAwsEnv() in
> this setup?
>
> And there is a precedent for environment variables for RocksDB Cloud
> configurations. They are used for credentials and AWS settings already.
>
> Environment variables are fine for credentials, I agree. They are not fine
> for an option that controls the path to which RocksDB Cloud syncs the data.
> I am happy to have hooks that lets you initialize config through
> environment variables, but they should not be invoked by default by the
> library.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rockset/rocksdb-cloud/pull/58?email_source=notifications&email_token=AABCTBDLWR7GCGOLE637JBTP4CNORA5CNFSM4HYDI4RKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYMO3KQ#issuecomment-504950186>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABCTBBVYZXVTLTEW6GICOLP4CNORANCNFSM4HYDI4RA>
> .
>

igorcanadi commented:
> The idea is that there will be somewhere in a RocksDB INI file a field that
says use "AWS" as your environment and that will case the NewAwsEnv to be
used.  There is an existing ObjectRegistry in RocksDB.  Imagine how you
would use it to initialize RocksDB Cloud.

I like it. Could this `AWS` environment also specify the src_bucket and dest_bucket?

mrambacher commented:
I am not sure exactly what your question is.

When I am done, we could dump any setting from the CloudEnvOptions into the
INI file and read it back in.  So if the source/dest bucket are defined as
configuration properties, then we can read them through the INI files.

I would recommend that we do not dump the credential settings for security
reasons, but every other setting should be possible (with almost no
additional code in RocksDB Cloud).  If you look at
https://github.com/facebook/rocksdb/pull/5362 and
https://github.com/facebook/rocksdb/pull/5361, you can see the start of
where this is going.  These PRs introduce a new Configurable class that
classes can extend that will allow them to be read/written from an INI file
and configured via APIs.  The next step (which I hope to PR later this
week), is integrating those into a "Customizable" class, which will
additionally allow these objects to be loaded via the ObjectRegistry.
Phase 3 is adding the ability to use Dynamic Libraries and the Object
Registry together.  I will be presenting this at the Meetup next month.

So I do not necessarily except folks to be configuring RocksDB Cloud
long-term via the environment variable settings.  It is a good way however
for bootstrapping it.  And at the moment it is the ONLY way we can test
RocksDB Cloud in our environment.  I am fine if you choose NOT to document
configuring RocksDB Cloud via the environment variables and it is a
"secret' way of getting it running.  But at the moment it is the only way
we can make any progress.

Note in a future PR we will be pushing a few additional environment
variables as well (around which type of AWS credentials to use).

On Mon, Jun 24, 2019 at 9:09 AM Igor Canadi <notifications@github.com>
wrote:

> The idea is that there will be somewhere in a RocksDB INI file a field that
> says use "AWS" as your environment and that will case the NewAwsEnv to be
> used. There is an existing ObjectRegistry in RocksDB. Imagine how you
> would use it to initialize RocksDB Cloud.
>
> I like it. Could this AWS environment also specify the src_bucket and
> dest_bucket?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rockset/rocksdb-cloud/pull/58?email_source=notifications&email_token=AABCTBGHTGLFAMYSAIQ3O43P4DBOZA5CNFSM4HYDI4RKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYM3UJQ#issuecomment-505002534>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABCTBEQY3JAOYTYNKA6GR3P4DBOZANCNFSM4HYDI4RA>
> .
>

igorcanadi commented:
> I am not sure exactly what your question is.

What I'm saying is there will be INI file where you configure `AWS` environment. That ini file should also contain src_bucket and dest_bucket properties. The code that reads the INI file should take the values from the file and populate `CloudEnvOptions`. That seems to be what you are saying as well, so I think we are in agreement.

> So I do not necessarily except folks to be configuring RocksDB Cloud
long-term via the environment variable settings.  It is a good way however
for bootstrapping it.  And at the moment it is the ONLY way we can test
RocksDB Cloud in our environment.  I am fine if you choose NOT to document
configuring RocksDB Cloud via the environment variables and it is a
"secret' way of getting it running.  But at the moment it is the only way
we can make any progress.

If this is for testing purposes until we can get INI files setup, I would rather not push it upstream but instead maintain in branch/fork. Would that work? We can push everything upstream, except the part where environment variables are read by default without a way to turn it off.

mrambacher commented:
I am not convinced that the environment variable options might not always
be necessary, if for nothing else as a way of bootstrapping the system, in
order to get to the configuration the first time.

If the values are not pushed into master, it will make our development very
difficult until we get to the point of being able to initialize everything
through configuration files.  And it will make the contributing code back
more challenging, as of figuring out how to manage the various branches and
changes.

I propose allowing the environment variables to go in.  If you prefer names
of the variables to suggest that they are for internal use or something
like that, I am fine with that.  If you choose not to document them, that
works for me as well.  If after the configuration/INI file stuff gets
merged it we agree there is no longer a need for the environment variables,
we can rip them back out.  Since they would be internal and not documented
in the interim, it would not be an API change and no one should be relying
on them.

At the rate changes are reviewed and taken into RocksDB and those changes
folded into RockDB cloud, it could take several months past when
development is complete before the configuration changes make an official
release.  It would be nice to be able to use the system in the mean time...

On Mon, Jun 24, 2019 at 9:29 AM Igor Canadi <notifications@github.com>
wrote:

> I am not sure exactly what your question is.
>
> What I'm saying is there will be INI file where you configure AWS
> environment. That ini file should also contain src_bucket and dest_bucket
> properties. The code that reads the INI file should take the values from
> the file and populate CloudEnvOptions. That seems to be what you are
> saying as well, so I think we are in agreement.
>
> So I do not necessarily except folks to be configuring RocksDB Cloud
> long-term via the environment variable settings. It is a good way however
> for bootstrapping it. And at the moment it is the ONLY way we can test
> RocksDB Cloud in our environment. I am fine if you choose NOT to document
> configuring RocksDB Cloud via the environment variables and it is a
> "secret' way of getting it running. But at the moment it is the only way
> we can make any progress.
>
> If this is for testing purposes until we can get INI files setup, I would
> rather not push it upstream but instead maintain in branch/fork. Would that
> work? We can push everything upstream, except the part where environment
> variables are read by default without a way to turn it off.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rockset/rocksdb-cloud/pull/58?email_source=notifications&email_token=AABCTBF7INC6U4PSU3CXSLLP4DD25A5CNFSM4HYDI4RKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYM5PEI#issuecomment-505010065>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABCTBHUYMYSQDFEMBF27O3P4DD25ANCNFSM4HYDI4RA>
> .
>

igorcanadi commented:
Before you have configuration support in RocksDB, where do you call `CloudEnv::NewAwsEnv()`?

mrambacher commented:
We are not writing a new product to use RocksDB Cloud.  We are trying to
get it integrated into existing products we are hoping to use (like
Rocksandra, Percona, MariaDB, etc).  We do not want to change those
products to use RocksDB Cloud, as it seems like an uphill battle.  However,
if we can be sure they can be configured from initialization files, we can
get RocksDB Cloud into them.

On Mon, Jun 24, 2019 at 10:53 AM Igor Canadi <notifications@github.com>
wrote:

> Before you have configuration support in RocksDB, where do you call
> CloudEnv::NewAwsEnv()?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rockset/rocksdb-cloud/pull/58?email_source=notifications&email_token=AABCTBCCRDAMHAQIYSQEJTDP4DNVZA5CNFSM4HYDI4RKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYNGDXQ#issuecomment-505045470>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABCTBB2SVSEBXIJSKQR323P4DNVZANCNFSM4HYDI4RA>
> .
>

igorcanadi commented:
@mrambacher my main point is this -- whatever you do, at some point you'll have to call `CloudEnv::NewAwsEnv()`. Your plan is to eventually call this from RocksDB based on some INI file config, good. Before this is done, you'll have to stick this call somewhere else (I don't know where). Now wherever you call `CloudEnv::NewAwsEnv()` from, replace this code with:

```
auto cloudEnvOptions = initializeCloudEnvFromEnvironmentVariables();
CloudEnv::NewAwsEnv(cloudEnvOptions);
```

And that's it. You can initialize whatever you want from environment variables that way, without changing the way we initialize `CloudEnvOptions` from within RocksDB-cloud.

mrambacher commented:
If we cannot initialize RocksDB Cloud from the environment, we cannot run
any RocksDB Cloud test without modifying the tests to call
initializeCloudEnvFromEnvironmentVariables.  If every tool that uses
RocksDB in the system does that work, then why bother having that call at
all and do it for them?

Until RocksDB adopts the configuration settings and RocksDB Cloud catches
up, we will have no way to run any RocksDB tests without doing it in some
custom fashion.  If we cannot run RocksDB Cloud tests, we cannot validate
any contributions we have to RocksDB Cloud back to the system.  In essence,
we will not be able to use RocksDB Cloud *anywhere* until RocksDB Cloud
catches up with this development.  At this point, we have a handful of PRs
queued up to contribute to RocksDB Cloud, the testing of which in our
environment all depend on being able to initialize RocksDB from the
environment.  We have added more credential support to RocksDB Cloud and
have separated the different engines (LogProvider, StorageProvider) to make
it easier to write additional Environment support.

Or are you suggesting we have a "test" method that initializes every cloud
test from the environment?

I am unclear as to what your objection here is.  We have not changed the
way anything will behave today but have added a potentially useful
configuration feature for others.  All existing code should continue to
work exactly as it does now.


On Mon, Jun 24, 2019 at 11:28 AM Igor Canadi <notifications@github.com>
wrote:

> @mrambacher <https://github.com/mrambacher> my main point is this --
> whatever you do, at some point you'll have to call CloudEnv::NewAwsEnv().
> Your plan is to eventually call this from RocksDB based on some INI file
> config, good. Before this is done, you'll have to stick this call somewhere
> else (I don't know where). Now wherever you call CloudEnv::NewAwsEnv()
> from, replace this code with:
>
> auto cloudEnvOptions = initializeCloudEnvFromEnvironmentVariables();
> CloudEnv::NewAwsEnv(cloudEnvOptions);
>
> And that's it. You can initialize whatever you want from environment
> variables that way, without changing the way we initialize CloudEnvOptions
> from within RocksDB-cloud.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rockset/rocksdb-cloud/pull/58?email_source=notifications&email_token=AABCTBDU473NICJFTUCHFHTP4DRYTA5CNFSM4HYDI4RKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYNJXPQ#issuecomment-505060286>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABCTBDO3EWYGX3NUS7BBBLP4DRYTANCNFSM4HYDI4RA>
> .
>

dhruba commented on cloud/cloud_env.cc:
```diff
@@ -15,7 +17,85 @@
 
 namespace rocksdb {
+  
+bool CloudEnvOptions::GetNameFromEnvironment(const char *name, const char *alt, std::string * result) {
```
Can we get code comments to describe the sematics of this method?

dhruba commented on cloud/cloud_env.cc:
```diff
@@ -15,7 +17,85 @@
+  }
+}
+  
+void BucketOptions::Initialize(const std::string & bucket,
```
I am not able to understand the relationship between Initialize, SetBucketName and GetNameFromEnvironment. Can you pl put in some code comments on which method is supposed to do what.  It will help us review the code.

mrambacher replied:
Added comments as requested

dhruba commented:
We still feel that it is better to introduce a new static method NewAwsEnvFromEnv() and leave NewAwsEnv() method as it is. This will ensure that the code path that  initializes from env variables is quite seperate from the code path that does not look at env variables.
```
static NewAwsEnvFromEnv() {
  // try reading env variables.
  if (env vriablea are not set) {
    return NewAwsEnv();
  }
 ...
}
```

Regarding the unit tests,  there are only two places where we call NewAwsEnv().  Replace it via the call to NewAwsEnvFromEnv().

igorcanadi commented:
@dhruba 's proposed solution would address my concerns.

mrambacher commented:
> We still feel that it is better to introduce a new static method NewAwsEnvFromEnv() and leave NewAwsEnv() method as it is. This will ensure that the code path that initializes from env variables is quite seperate from the code path that does not look at env variables.
> 
> ```
> static NewAwsEnvFromEnv() {
>   // try reading env variables.
>   if (env vriablea are not set) {
>     return NewAwsEnv();
>   }
>  ...
> }
> ```
> 
> Regarding the unit tests, there are only two places where we call NewAwsEnv(). Replace it via the call to NewAwsEnvFromEnv().

I am not sure of what you are suggesting and how it works and is not a source of greater confusion.  

First, the properties set in the environment are embedded in CloudEnvOptions.  So the NewAwsEnvFromEnv would need to update the properties in CloudEnvOptions?  Or should there be a CloudEnvOptions::InitializeFromEnvironment?

Second, what are the semantics of this "InitializeFromEnvironment method?  Does it override the values if they are not set or always override them or what?  What do we do about the "rockset." prefix option, which is set by default?

As it stands now, the environment options are the initial values (if they are set).  If someone calls "initialize", the values are overridden if the arguments to initialize are set and ignored if they are not set.  If someone calls the "set" methods, the environment variables are ignored.  There are a set of tests in db_cloud_test under BucketTest that do exactly this sort of logic.

So, what is the concern and how do we fix this stalemate?

mrambacher commented:
I think I addressed this issue in the latest push.  There is a TEST_Initialize method off both CloudEnvOptions and BucketOptions.  The BucketOptions one will set the bucket properties based on the environment variables.  The CloudEnvOptions calls the BucketOptions one, updating both buckets.

I rebased to collapse the BucketOptions into a single commit.  There is still an earlier commit for the BucketName et al, methods.

igorcanadi commented:
The final change looks good to me. Would you mind rebasing to resolve the conflicts?

@dhruba Do you also want to take a look?

dhruba commented:
@mrambacher : can you please submit this as one commit (rather than 17 commits). Also, pl merge with latest master so that we can pull it into master?

mrambacher commented:
I rebased onto the latest master (this resulted in an additional PR to
address build issues under MacOS).  I also squashed the changes into a
single commit.

Please let me know if there is anything else.

On Mon, Jul 1, 2019 at 3:37 PM dhruba borthakur <notifications@github.com>
wrote:

> @mrambacher <https://github.com/mrambacher> : can you please submit this
> as one commit (rather than 17 commits). Also, pl merge with latest master
> so that we can pull it into master?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rockset/rocksdb-cloud/pull/58?email_source=notifications&email_token=AABCTBDJ7LGNSW3O7BNEMFLP5JMHJA5CNFSM4HYDI4RKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY7EEMI#issuecomment-507396657>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABCTBADVXO6I2AIEPLJKTLP5JMHJANCNFSM4HYDI4RA>
> .
>

igorcanadi commented on purge.cc:
```diff
@@ -0,0 +1,267 @@
+// Copyright (c) 2017 Rockset.
```
What's up with this file? An accident?

mrambacher replied:
I do not see this change

igorcanadi replied:
What do you mean? There is an extra `purge.cc` file in the commit, isn't there?

igorcanadi commented on include/rocksdb/cloud/cloud_env_options.h:
```diff
@@ -296,14 +339,18 @@ class CloudEnv : public Env {
                           const std::string& dest_bucket_name,
                           const std::string& dest_object_prefix,
                           const std::string& dest_bucket_region,
+                          CloudEnvOptions& env_options,
```
Can we keep this as `const CloudEnvOptions& env_options` and make the copy inside of `NewAwsEnv`? Otherwise this is a very subtle public API breaking change. One copy won't hurt.

igorcanadi commented on include/rocksdb/cloud/cloud_env_options.h:
```diff
@@ -60,12 +60,48 @@ enum class CloudRequestOpType {
+  void SetRegion(const std::string & region)  { region_ = region; }
+
+  // 
+  void TEST_Initialize(const std::string & name_prefix,
```
Here and everywhere else: our code style doesn't put whitespace between `&` (or `*`) and the type name.

So: `const std::string& name_prefix`, *not* `const std::string & name_prefix`

igorcanadi commented on cloud/cloud_env.cc:
```diff
@@ -143,37 +211,48 @@ void CloudEnvImpl::TEST_InitEmptyCloudManifest() {
                            const std::string& dest_cloud_region,
+                           CloudEnvOptions& options,
+                           const std::shared_ptr<Logger> & logger, CloudEnv** cenv) {
+  if (! src_cloud_bucket.empty()) options.src_bucket.SetBucketName(src_cloud_bucket);
```
No space between unary operator and its operand.

igorcanadi commented on include/rocksdb/cloud/cloud_env_options.h:
```diff
@@ -60,12 +60,48 @@ enum class CloudRequestOpType {
+  std::string prefix_;
+  std::string object_;
+  std::string region_;
+  std::string name_;
```
can you comment on all the various options? `bucket_`, `prefix_` and `name_` are fairly confusing. Also, mention that default value of `prefix_` is `rockset`.

igorcanadi commented on cloud/db_cloud_test.cc:
```diff
@@ -373,7 +368,7 @@ TEST_F(CloudTest, Newdb) {
     std::unique_ptr<CloudEnv> cloud_env;
     std::unique_ptr<DBCloud> cloud_db;
-    CloneDB("newdb2", src_bucket_prefix_, src_object_prefix_, "", "", &cloud_db,
+    CloneDB("newdb2", "", "", &cloud_db,
```
Is this correct? Previously we supplied `src_bucket_prefix_` and `src_object_prefix_` here and now it's empty strings.

mrambacher replied:
The src_XXX options were removed, as they were always being passed in as the old values.

mrambacher commented:
Sorry, yes there was.  I missed the directory portion of the comment.
Cleaned up now.

On Tue, Jul 2, 2019 at 1:53 PM Igor Canadi <notifications@github.com> wrote:

> *@igorcanadi* commented on this pull request.
> ------------------------------
>
> In purge.cc
> <https://github.com/rockset/rocksdb-cloud/pull/58#discussion_r299609114>:
>
> > @@ -0,0 +1,267 @@
> +// Copyright (c) 2017 Rockset.
>
> What do you mean? There is an extra purge.cc file in the commit, isn't
> there?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rockset/rocksdb-cloud/pull/58?email_source=notifications&email_token=AABCTBHSJTCD7YK4A6L2F63P5OIYPA5CNFSM4HYDI4RKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5JFLZA#discussion_r299609114>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABCTBD4GI557CNEO2CO7ATP5OIYPANCNFSM4HYDI4RA>
> .
>

igorcanadi commented:
Ok, lgtm, let's merge this!

ravibiz commented:
Hi, In one of our use cases we are using rocksdb as embedded database and trying to clone local db directory to aws(s3) in regular interval to achieve durability.  We are unable to run sample test and got confused with these naming's in the code 'bucket', 'prefix', 'suffix' and src/dest buckets. I may be misunderstanding here, need your help on understanding the same.

Local db directory = /tmp/rocdksdb_sample  (would like to make it durable, cloning to aws s3)
S3 path = s3://rocksdb/payments/data

Which signature I should use to create awsenv instance?

Thanks
